File: leetcode/53.最大子数组和.py
class Solution:
    def maxSubArray(self, nums: List[int]) -> int:
        # 简单dp
        n = len(nums)
        f = [0] * n
        f[0] = nums[0]
        for i in range(1, n):
            f[i] = max(f[i - 1] + nums[i], nums[i])
        return max(f)

File: leetcode/53.最大子数组和.go
func maxSubArray(nums []int) int {
    // 动态规划
    // dp[i] = max(dp[i - 1], dp[i - 1] + nums[i])
    // return max(dp)
    n := len(nums)
    f := make([]int, n)
    ans := nums[0]
    f[0] = nums[0]
    for i := 1; i < n; i ++ {
        f[i] = max(nums[i], f[i - 1] + nums[i])
        ans = max(ans, f[i])
    }
    return ans
}